src/main/java/org/saltyrtc/tasks/webrtc/WebRTCTask.java

```diff
@@ -502,6 +509,7 @@ public void close(int reason) {
             this.sdc.close();
             this.sdc.dispose();
             this.sdc = null;
+            this.signaling.setState(SignalingState.CLOSED);
```

Comment:
Does that mean the task never went into the closed state? And why in the conditional?

Reply:
Just read the description. :smirk_cat: Okay, I guess this makes sense but should probably not be inside of the conditional.

Reply:
We don't know, this is just making sure that the signaling state is set to CLOSED in case that hasn't been done yet. (The saltyrtc-client implementation ignores repeated state changes.)

I'm not sure why this used to work previously but doesn't work anymore, must be a change in the event call order in the WebRTC native library.

These are the data channel events logged when closing a Threema Web connection from the webclient or from the app:

> 2018-11-12 17:30:51.242 22495-29251/ch.threema.app I/SaltyRTC: DataChannel: State changed to OPEN
> 2018-11-12 17:31:01.794 22495-29251/ch.threema.app I/SaltyRTC: DataChannel: State changed to CLOSING

Seems that the onStateChange() call for the CLOSED event is done *after* setting the data channel to null.

I thought about enclosing a reference to the data channel inside the event handler, but that seems a bit dirty to me, since we'd be updating states from unreachable objects.

Reply:
> Seems that the onStateChange() call for the CLOSED event is done after setting the data channel to null.

If so, that would indicate a bug in the referencing counting of the implementation.

Reply:
What we could do alternatively is creating our own `DataChannel.Observer` like interface that includes the new state as parameter when being called. This should work because in the `SecureDataChannel` the data channel reference is never cleared, it is simply left to the garbage collector (since we treat the `SecureDataChannel` as a transparent proxy to the actual `DataChannel` instance).

It would mean that the SDC isn't a drop-in replacement anymore though.

Reply:
> It would mean that the SDC isn't a drop-in replacement anymore though.

Don't forget that this wrapper is going to be removed soon anyway, so I wouldn't put too much effort into it. :slightly_smiling_face:

I still don't understand what the problem is to simply unbind all events. Is there no *unregister* method?

Reply:
Ah, now I get it, you want to unregister the event from the `close` method, before closing, not from with in this handler, right? (Since in this handler we don't have a reference to the sdc anymore.) But then you miss the `CLOSING` event.

Reply:
> If you unbind the events after closing that might not work as intended in case the events are handled asynchronously.

That's fine since you set it to `CLOSED` anyway.

Reply:
> Okay, I guess this makes sense but should probably not be inside of the conditional (unless sdc being null is just a safety check)

It is just a safety check, in case the user would call `close()` twice in a row. I don't think we should set the state to CLOSED when not actually closing something.